puzzle = "{{{{{{<!!u!aa<e!!!a!a!>,<o!!!!o>,<oio{>},{{{<,!>o!>},<!>},<!i!{,a!!}!!!>!>},<!!aa>,<!!!!!>>},{{<a!!!>a!>},<!>},<!,!>},<!>eua>},<!!!>ua!>!>!ua,!}u!,ieuau{{a>},{{},{{{<a!!o,o<>,<e!>},<!>iiu!>,<!>{!>,<>},{{<!>,<o,!!!!u>}}}},{{<u!a!>!!!>},<!!!>e!!!>!!a!!o<i>},{{}}}}},{},{{{{<!>,<o!>{oi!>,<>,{{<u!iae}{!!aia>},{}}},{<!>,!o>,<a>}},{<!!!!!>o!>},<{>,{<!o!>},<aa!>},<uu>}},{{<}i!!>},<}!>},<!!!!!>},!!!>},<}a!!!!!>!!u>}},{{<!!o!!!o>}},{<oue{!!!!{a,,!!o>,<a!>i,!!!>!!uaa!!<>}}},{{{{}},{{{<!>{!!aie!>!{!o{!!}u<}aa>},<a,a}<aa!>},<>}}},{}}},{}},{{{{{<au>}},{{},{{<!>>},{<!>,<a!!!>!>},<!a!i!>},<!>},<!!ie!!u>}}},{{{<!>{o,e!!!!!!!>,<a!!!>e>}}}},{{},{<!>},<oaaa!!!!u!!!>!!}oe>}},{<!>},<!a!>,<!!!!!>!>},<!>,<!>},<{!{!!!>!!<}!>>}}},{{{<o{i!!!a!>a!>,<,!{!>,<!!ua!>a!>!>!>},<a>},{<!!i<!!!>}aiua<,!a{!>,<>}},{{{},{{{{{<!!!>},<!!<!>a,!>,<!>,<!u!!!>>,{{<!!!>>}}}},{}}},{{{},{{{{{}},<!>,<!!!>i!!,!{!>,<!!!><!!ae!,iua>},{{<!!e!u!>!au!>,<!!!>!!,!!e,!!!>!!ui}aa>},<!a,>},{<!>,<}a{!!{!!e!>!!!!>,<>}}},{{{}},{{<,ee!a},>},{{{{{{{}},{<!!u}!!>}}},{}},{{},{<!}!!,!>,o!!!!a>}},{<!>,<ee!>i!!!!!>u,,,>,<>}},{<!a<!>,<a!>!!!>!!a!>,<a!!au!!<!!!!!>!>},<<>}}}},{{<e!><!!,<aa!!>,<e!<}}>}}},{{<!!!>u!>,<io!<i<i>,<!!!>!!!!!>!!!>>}},{{{{<!e!!aiaa!!!>},<!!<}>},{<!!!>}{au!!{!>,iu!!!>o>}}},{{{},{<ui!>,<!>},<!!a}}!>,<!>,<},au!>i!!>}},{{<e>},{{<<!!!!!>}!>,<u{a!!{i>,{{},<!!!>},<{!>},<aa!>},<<!>,<u>}},{<!!!>!,!>},<a,!!ee!>,<!!},,ao<o>}}},{<}{au}!>},<u!!!>!,ai!}!>,<!!!>,!>,<{i!>,<>,<<>}}}},{{{{<!!}!a!!a}{!ioi>}},{},{{{}},{<au!>!>},<!a!!!>ie!>},<e!!a<,}>}}},{{<!!,}>,{{<!a!>!>,<!>!!{iaa{!>,<}!!!>!!,e!!i!!!>>}}}}}},{{{},{<!><!}i!!!!u<!!!!a!!!!!!u!>,<,!a>}},{{},{<}u}!>}!!!>i!!a!!eae{!!!>},<>,{<!}}}!>},<a!>,!>},<o,,e!>,<!uu>,{{<!>e,!>},<,u!a!!!>o!!a<!!!>!!!>>},{}}}}},{{{},<!!!>},<!>,<o!>{{!>,<}ao!>,<uo>},{<!!!!,o}<!!,!>,<>,<,ue!!{!>},<}>}}}},{{{<!>,<{!>}o>},{<e!!!!!!!!!>!>,<!!a>}},{<,a!>,<!{}oaaa>},{{}}},{}},{{{{<!>,<,{!<{,!>i<!>,<!!<!iu}!>},<i!}>},<!>,<aaa!>a!!!>!>},<!>>},{{{{<e!!{!>},<>},<o!!u!>},<a!,!o!>i}{!!a!!!!!>!e!!,!>},<!!!>,<>},{{{<{aai>}}}},<!!a}>},{{<i!>e!!!!o!>a}ae!>,<!!i}!!<u>},<>}},{{<uoei!{!!!i{ei!!!>!>,<,!a!}!>},<!!e>}}},{{{<a!i<aa}a!>},<<o!!!>!!!>,<!!!>>}},{{{{<!>,<uo!,,,!!!>!!!>!>!!!!!!oa>},{}}},<!!!!a!!!>!!!!{!>,<a!>},<!aaa!!>}}},{{{{},{{{{<aa}!!u<>,{<>}},{{{<o!!!!,a<!!!>!>},<a!o>},{}},<}aa!a>},{{<,u!oao}!}a!!{a!a!!!>!!u>},<i,>}},{{{<!>,<a,!!!>!!!>u}!>},<}!!o>},{{}}},{<}<uu>}}},{},{{},{},{{<ui!!!>!!!>!!<u!<<i!>,<}!!u!!!>ou!!!!a>}}},{{{<!!!>!a!!!>,<!!}!o!!<a<ia,u>,{<aa}u<e!!!{!!!!!>,!>},<<a>}},{{{{<<,!io!>,<u!>,<{!uu!!!>!>},<!oio!!<u!!a>}},<ae{}<!!!>!!!>!iaa!>},<!!iou>}},{<!!,!!!>!!!>eaa!>},<a!!!!!!u,>,<!!!ea!a<!!o!>},<a!>,<!>!>},<o}ea>}},{{{},{<iiaoa,!{,a!!!!!>},<>,<!>,<u!!o}<!>o!>,<!>},<,>}},{<!!!!!>,<i!!!!{,>}},{},{{{{<{{!>,<!>},<a!>,<>,<}!!{!>,<o!!e!>,<u<>},{<i!!!>,<o!>e!!!>>},{{{},<}!!{!!i!oa!!a!>a!,a!!<a!!a>},{}}}},{{}},{{{<>},{<o!>,<!au}!!>}},{{},{<!!<aae}a!!<ua>}}}}}}},{},{{{},{<a,}!!!>!>},<!<{i{!>},<i!>!>,<!!a}a!!u{!>>}},{<a}!>,<,a<u!!oe!!!>,<>}}},{{{<}!!!>!>,<!>},<i<!>,<au!!!>},<a>,{<i>}},{{<!!a}!>>}},{{<>,<,a!>},<uae!>!>,<a!!!>a>},<!!e!!ua<}!}o!>},<oe!u!!!a>}},{{},{<!><o!!!!!><!e<a!>,<i!u<!>},<o>,<{!!!>,!!e!!!!e!>!!!>aoe!!au{i>}},{{<a!>},<!!!!ua>,{}},{{{<o,>,{<!!!!<>}},{<,>}}}}},{{}},{{<!ae,>},{{},{{{<},a!a!>},<,aoi>},{<!a!>},<a!>},<!>},<a!{!>,<ua!i!u!!a!oi>}},{<!!a!!!a!>,<!>,<}!>},<!>},<!!{e!!!>,<>}}},{<,o,}o{!>},<,!!!>},<oo,e>,{<!>u!>{}}}!>,<i!!!>a>}}}}},{{{{{<!!!!!!!>e!!<!!!!!>!a!>!!!>a!!aaa{!>},<u>},{<!!!!!>!!!>,ao!>},<!!aoi>,{{<!!oa},ia,ua!!a<a{!,!>!!e>},{<!!!!!>,<!}o{,>}}}}}},{{{{{{{<!>},<a!!aa!!o,!>!!!!uo!!!!!>!!!>>},{<<ua{!!!>},<i>}},{{<u!>e!!!>!>a!!<!!!>ee!>},<!!!>{!!!>io!>,>},<}>}}},{<!!!>}aa{a!!uu!>>}},{},{<!!,!!e}o}!>}!>,<>,{<a,!>},<ou}>,{}}}},{<iu>,<>},{<!>},<!!oo!>i}>}},{{{{},{}},{},{{{<}!!!!!>!,<i,!>u!!!!}{a!o!!!>,>,<o!!!>!!,!io>},{{<a!>!!i!a!!a!>}a!!!>},<>},<{a>},{<a!!,,!!a!>},<!>,<{a,!!>,<!>e!>},<a!!!>!!<u!>},<!>,<!>{!>},<!!i!>,<!iie>}},{{{{}}}}}},{{<!>}>}},{{{{<a!!!!!>{!!!>},<o>},<>},{<oua{!i!>>,{<ai,eu!!!>au!!!!!>!!!>a{,!>,<aai>}},{{{}},{<!>},<!>aaa!a!!a>,{<!>,<{u!>},<,ia!!!o!a{i>}},{{{<!!ee!<!!!>!!!>o>}},{<{iioa>}}}}},{{{{},{<!!!>a!!!!e!>},<e!>ea}!!}a!!>}}},{{<>},{{<}!!}!>!>,}i}!!!>a,!!!>a!a!ai>}},{<!!,>,<!>,a!!!>,<!>},<!>,<!>},<e,{>}},{<,!>,<a!a!!!!!>},<u{!>},<ao>}}}},{{{{},<e!i!>,<<}a!>!!a>},{{<<!!iu!>},<a!>},<eu!iu!!!a!>},<o!>,<a!!,o>,<<!>},<!!!>a>},{{<,!>e<>}},{{},{{}},{<!>!>!!i!>>}}}},{{<!>!>ai!>,<!!!>>,{{<o!!!!,!!!!!!!!!><!>},<{!>eu!>},<!!!>!!<i!,!o>},<!>},<o!>!o!!,}!>},<!!a!!!>!}e!!oa!aau!,>}},{},{<}a!>,<!!!>!!!>!e}!>!!e{}aa>,{<!>ae!>},<u!!!>o!>},<!{!>,<!!u<a!>},<>}}}},{{},{{{<}!>},<!!oa!!i!>,<!>},<!>,<!!!><u!>,<a>,{}}},{{<!>!!,<,a,!>},<>},{}},{<<!!!>}u!!!!!>},>,{{<!!a!aa>,{{{<iu!!!!!>},<!!!>>}}}},<>}}},{},{{}}}},{{{{{{{},{<a{a{!>,<!!!>!!!>uauao{a!!!>u,i!!!>>}}}},{{<a,!!!>},<a>},{{{<!>!>},<>},{<!>,<!><!>},<!!{a<uoi!>,<e>,{}}},{{}}},{<!>,<>,<!>},<aeae!!a<!>,<!>!>}}!>,<!>},<u>}},{<}i!!!>},<u>}},{{<!}!>o!!!{}!!!>>,{}}},{},{{{{{<>}},{{<!<,a!oa!>},<!>},!!!><!>},<>},{<!i!>o!!!>},<!>},<e<{e!>!<!>},<a>}},{<a<,!aa!>{a,u,}u>}}}}},{{{{{{{<u!!!>o!!,e!>!!ai!!!aa!>,<}!>,<!!!{>}},{<!!o>}}},{<u!>!>,<}a!!>}}},{{{<ui>,<!!!>eu,!!!>oi!>,<!!!!!>>},{{<!!aa!>},<a!!!>!!,{}!!ai!>a>},<!>},<!!!>eoai!!o>},{{{<!>},<ui!>!>,<ie!>},<uu>},<o!u!>a!>},<e!>ae,<e!>,<io!>},<ai>}}},{{<!uo!!!aa!!!>},<a!!!>o<aea!!aeo,<!!!>>,{<aua!>},<!!<!>,<,i>,<!!!>a!!!}e{!!!!{<eu!>},<!!!>!>},<,!!aa!>>}},{{{{<!!!>u}a!>,<ea>},<!!>}}},{{{{}},{<!>,<!!ao!!ae>}},{<!!a>}}},{{<!>,<ie!!!>}<i!<!>},<a>,{<},!,ao!>,<ea!!!>a!o!>,<a>}}},{{{{<a!!}!!a}!!oo>,{<!>,<{,ao<aa!!a{>}},<{aaaa}!>},<<!>},<<!!a!!>}}}}},{{{{{},{{},{<>}},{{<!!!>},<!>},<a{aa{eo>},<!><!!uiaa,a<aa!>},<!>,<<{ao>}},{<oa{,!!!!!>,a>,{{<a{<i}e!!!>{>}}},{<!>ae!!e>}},{{{<!!!>,<},ae<!>!!!>ae!!!>>}}},{{<>,{{{}}}},{{<!e!>,<a<aa!!i>},<a{!>!>!!!>{o,!}o!>},<o!>,<!!!}a!!>}}},{{{}}},{{<a!>,>}},{{{<!>},<!!!u!!o,!!aa!!!aea!!!>a>,{<!!{{{!>!}!!oeu!>u!!e}a!>,<e!>,<,>}},{{<!>},<ea}!>u>},<oai!!!>!!o!uu!>},<,ae>}},{{<e!!!>!>,<<aa,u,>},{{<>}},{{},{{<a}!>!>}!!!>!!!>!<u>},<!!!>a!>,<ae!>i}!!!>oo,!a!ue>}}},{{{<>,{{},<}!!<>}},{<>}},{{},{}}},{{},{}}}},{{{<!>},<oiuu!{!>,<!!{>,{<oa!!<!>,<i,!>!>,<>,{<{!!!!a!>},<a>}}}}},{{{{<<!!!>!!!>>}},{{<u!>,<i!{a>},<!!!>oa!>,<!>},<<!!!>}>},{{<!!iie>},{<!{!!e}ai{!>}!ou>}}}}},{{},{{{{{},{<!!o<i!}!}ee!!!>!>},<a!!!>},<a{ae>}}},{{{<o}>}},{},{{{<!>,<!!e!>,<>}},{{},{<!e<!!!!}u!>,<,ui!>},<i,u}!a>,<>},{{<a!!<}!>u!!i!!!>,!>,<eaioa<u>}}}}},{{}},{{{},{{}}},{{{{{},{{<u!>},<!>,<i!>!>,<aa<>}}},{<aa>}}},{{{<u!!a!!a!>!>},<i{u!>},<a>}}},{{},{<<ei!{!>},<>}}},{{},<,!!!>!!!!!>,<!!{a}!!!>,<!!!a!!a!!i!!!>i!!,,}{!!>}}},{{}},{{<!!i!!!>e!!!>,<!}!!!>!>,<e!!o,,oaa>,<{o!!!!!>i{!>>}}},{{{<u!!}aauaa<}o{!a>,{{},{<!>},<oo!>,<i,!>u!>u!a!!aae!!!>,<i!!!>>}}},{}},{{{}},{{<!!i<!>,<!!!ao!>},<!!!!!>,a!!!!!<>},<>}},{{{<!>,<!!,!!a!>},<!!!>},<!!u!>!><!a!!!>}!!oe!!aa>,{}},{{<!a{aaa<!>,<>,{}},{}}},{},{{},{<ao!!!>!>!!a,e!!!>a!!!>!!!>u,a}>}}},{{{{{}},{<!!!>,<i,!>,<ei<au<<!!o}>}}},{{{}},{{<!aa!>},<>},<uaa!!!>a!>},<!!!>a!oau>},{{<,!!!>!!!!!>,<{ua!>,<a!>a!!!>,<a!!<{>},{}}},{{<e!aa!!!>ei{oa<!!o!>,<au>}}}},{{{{{<!!{}!!!>!>!>,<!>>},{<a!>o!>,<!!!!>}}},{{<!,!!!>},<i!>e}!!<a!>,<e!!o!!<aio>}}},{},{},{{{},{}},{{{<,,!!!>{>}},{{<!>},<i<e}!!ou<!>e>}}}}},{{{<,!>},<!<<<!>,<!>,<}!!e!!!>,<a!!!a<}!}!ia>,{{<i!>!!!>{>}}}},{{{{{<!!ae>},{{<>}}},{{{<a!!!>,<!!!>i>},<ia}!!>},{{},{{{<a<!!a!!!>,<!>,<!<!!{}!>!!{ao!!>},{<uauaaaoa!!}<!>,i{a!}>}},{{<e!!,!>,<!>!!!!!>,<u!!!>aei}a>},<{aa>},{{<a!!!>i}i!>},<a!!!>,!>!>,<!>,<>}}}},{{<!,e!!!>oeaia}!!a!!!>u!>!!>},<!!!!!!}!!!>!!!!!!!>a!>,<,!ouoae!!!!!a!!!>,o{o>}}},{{<}o!!,!!{aoae>},{<!!!>a!!!>!>},<!>},<>}},{{<i!!aa>}},{{{<<}{!>},<!!!!!>ao!!a!e!!a!>,<o!>,<!>,<!!a!>,<aa>},{{{<<aa!!>},<}{e}<aiu{!>iae!>},<>},{{{<>},<!>,<a!!ue!!u}<{!!!>{,}!!!>!!!!,!>},<!>,<<<>},<e!>},<!!!!{!!!>!<e>}},{{{<a!!!>a!>!}!!eeo!!ui!>>}},<!>,<!>},<>}},{<!!a{>,<!,}{!eaua!>,<!!a!!!>u!>,<a}!ee}>},{<,a!!u>}}},{{{{{{{{},<!!!o,u<a{a!>!>,<!!!>a!o!!a<o}!!!>!!!>>},{<!!!>ii!!ie!!!>!!!>},<ua!!aei!au>}},{<!aoi!!e!!!!!>e}u!>},<!>},<!!!>!!!>{aoa!!e>,{<oo{!{i!!!!!>!>},<e!>>}},{<!o,>,{}}},{{<eao<>},{<!>,<!{a}>,{<i>}}}},{{{{{<!!oo!!!!,!!}oa<}u>}},{{<!<ae}!!!>!!!>!>o!!!!!!<e{e>}}},{{{}}},{}},{{<{e!>ie!>,i!>},<a<<!!aie!!>},{<!!,!aa!>!!>}},{{{}},<oo!!!>o<!!!>>}}},{{{{<!>a<!!!>e!!!>,<a!>>,<!!!!a!>!>},<!!eaao!!!>}>},{{}},{{{{<!>,<!>},<a!!},>}},{},{{{<{ioo!>,<a,a!!a!>},<!>},<!!!!u>,{<!>},<>}},{<!!!>,<oe!a!>!>},<}aia}!{,a!!!>>}},<io{aui!>a}!>},<ae>}},{{<a!>!>{<!>,<oa}i!>!>},<<>,{<!>,a!>},<!i!!!e!!!>!!!>!>,<!>{>}}},{{<>,<{e!>e!>{,{!!!>e{aa>},{<!!!!!><}!!!>aiaaa>,<!>,<!!!!eoee!!!>a,eoe>}}}},{{{<!ao!!ee!!a>},{<},<!>,<u!>},<u!!!>>,{<!!!!!>!!!>i!!!>!i,a{o>}}}}},{{{{{<>},{<!>!>},<aoa!!i!>},<!>,<!!!!}!!!>},<!!!>!!,>}}},{}},{<>},{<<!!!>!!a<a>}},{{},{{<o!>eea!!!>!>o!>{u!>!>,<ai!>},<!u!>!!>,{{}}},<!u{{}!!!!ae,e!>},<>},{{<!!!!},<!!!>{!!i!!<!>,<>,<!>ae!>},<u>},{},{<<!>},<e!>!!!>,!>,<a{a!>,<ai!!!>a!>!!>}}},{{},{<!!!>,<>},{{{<!!!>!!!>,<!oeio{i{e<!>,<e!!!>!!!>a!>o!!!>,<>},{<!!,a{}!!!>a!>a!>},<>}},<!!u!!!>!>,<o!}>}}},{{{<<!>},<ai!!!!!i!>},<!!<!u!>},<<a!>aai!>,<!>,<>},{},{<eu!>!>,<!!!>,<!!!>}}i!!!>},<>,<,<!>},<{!uae!!!!i!ii!!!>!>,<oi>}},{{{<>},{<<a}!>,<a{,a!>!}!>!!!>!!!!>}}}}},{{{},{<a!>,<!!u!>!{!>,<>}},{{},<{<!}!!!>!!e!!e!>!>a!{,,>}},{{{<!>!!<<!!,!!!>,}!>},<!>},<uu!>,<!!>},{{<aa!>!>},<<!!a}!a>}}},{<e!>,<}}!!oaau!>>},{<!!}!>,<e!!a!!!!aeu>,{<,!>,<!>,}}!>,<!>},<!!!>!!u>}}}},{{{{{},{{<{eaou!!o!>oa{i>},{<aa!>},<!!!>!>,<i!!,<i<!>},<i>}}},{<!>,<}!!!>!>},<!>{e!!<!!!>}!e>,<}!!!!>},{{},<a,oa{>}},{{<o!!!>{a!>!>,<}}a{}ia!oi!>},<,!!i>,{<,!!!>,<{!uei{>}},{{<uau!>,<!!!><>}},{{{},<>}}},{{{{<u!!!>},<!!o!>,<u!!!>a!!e!!!>!>},<e,!>,<,!!!>!>!!ei>,<!>,<!>i!!!>!!o<aa{!!,!>},<ea}!>{>},{}},{{{<!>!>,<!>,<oui>},<!>},<>},<!u!!a<!!,a!!!!!!o!>a!>!>!>},<>},{{{}},{<a!!}!!a!!!>!!u!!!>a>}}},{{{},<<{},!>!>},<>},{{{<a!!!!!>,<}<{}e!!!!!>!>,<>},{{{<!!<},!!!>!!!>{!>,<!>,<i>}},{{<a>,{{<u!>},<a!!!!!>!!!}!a>}}},{}}},{<>}},{{{},<a,i!>},<o!!!>!>,<!!!!aaua!>!!!u!<>},{{{{<ea!!!io!!u!!!a>},<e!!aa>}}}}},{{{},{{{<!!!>!!!!!>>},{<{,!{{}{u!!!!!><{oua>}},<a!>!!!>u}!>!>,<a!!oa!>>}},{<!<>}}},{{},{{{<!>},<!,!>!!!!aa>},<aoa,!<a!>,<!!,!>},<a<!!!>!!!!,<>}}}},{{{{{{}}},{},{{<a>},{{<!!!!}!><o!!!>!>,<>},{}},{{{<uaa!>!>!>},<!!!>ui!ao!{!!ai!!!>>},{{}},{{{},{{<!!<!>a!!oa!u!!!>!!!>e}u!o>}}},{}}},{{},{<u!!a!>},<}iu}!{a!>},<!ua>}},{<i!>a}{}}<a!!!>!>!!!>{>}}}},{},{{{},{<a<!a!!a!>,<a{u!!!>!i>}},<!>,<iu!>!>!!aa<a!>},<!a!!a>}}}},{{{<ua!!!>},<!>},<!!i!!<!>,<!!!>!>,<!!i!!!!<a!>},<>}},{<!>!>,<<!!>,<}e<!!!!ei!,,oa!!u!>,<!>,<ae!>,<u>},{<<o!>a<!!u!!!!!>},<{oe!>,<<>,<!!!!!>},<a!a{!>,<}!>},<u!!<u!>},<!!!>!>},<!!i{!>,<oe>}},{{}},{{{{{<!oe!!!>a!>},<<,>},{}},{}},{{},{{{},{<a!!ao!!!!a>}},{},{{<!>,<>}}},{<!!!>},<!!a{a!!}{o!}i!>},<{!!!>},<a!>},<u,>,<e!!!>o!o!!!!,!>,<!!aeu!!!!!>},<!>,<a}>}}},{<!!a!{!i!>},<au!<ua!>>,<!>!>!!!>ao!!o!!ua!,{>},{<<!aa!!!!!>i!>},<a!!e>,{<>}}}},{{<!!!>,<!!!!!!!>,<!>!>,<{!>,<!ue!!<!}a!!!!a!>o>},{{<!<!!!>aa,!>!!!>!!!>aaao,!<}!!!>e>}}}},{{<!,!!!i,<i}}!>},<a,!!!<!!>,<{!,a!!ao>},{{<!!!>u>}},{{},{<i!>!!a!!a<>}}}}},{{{{},{{{<}!>},<!!!>a!!e!>},<o}!>,<e>}}}},{{{<!>},<{ua!!!>uu!a!ii!!!!a!!!>!>o!>},<>},{}},{{{{<!!!!!!!>!>!>},<<!>,<!>},<e>}}}},{{{<>}},{<uu{{a!!!>!!aueo!!e!!!!!!ua>}}},{{{<!!!!!!!>>},<<>},{<e}>}},{{{{{<o<ao,!>!>},<<!>!>},<<!>a!!,o>},<i<},,!>,<a!!ia!aea>},{<!>,<u!>,<!>,<a>}}},{{<<a!>,<}!>,!!!>!>,<<!>{>,<a<}!!}!>,<}!!e!>!!a!!}!!!!>},{},{{{{{<e!!!>!>a!>!!!>aa!>},<!>},<!!a!u{!!!!o!!a!>!>>},<{!ou!!!>},<!>}a>},<!a!,a!!o!>,<a>},{{{},{}},<o!>!!e!!a,i}}!!{!!!>}<uo>},{{{<!!aa>}},{{<a<!!!>,<!!!>}{<!!!>!ea!!!a>},<!!!>},<!>i!>{!>},<a!!!>}!!i>}}}}}}},{{{{<{aeu!!,a!!!!a!>!{oi!!>}},{{{<>},{}}}},{{<!ia!>,<>,{<aa>,{<!,ie>,{<!!!>!!{eu!>,<,ea<>}}}},{{<!>,<!>,<!>},<a{o!!{!eaoeu<a},eou>}},{<,!>e>,<!>},<a!ia<e<ii}<!!>}},{{{<a<!>},<!}<!!ao,!>,<!i{iueaa!e>}},{},{{},{{{{{}},{<!>aaa{u<!!oa!!!!!>!!!>!!!>a!!!!!>},<u!>o>}}},{{}}},{{{{{},{{{{<!>!!!!e>}}},{{<!!!>!!!!{}!!!>},<}>}}},{{<!!!!aa!>},<!>},<}{!!oua}!>},<!!!euu}!>,<>},<!>,<!!!!!>ue!!ai!>},<{!>!>},<{ai,!a!!!!!{!!!>!>>}},{<a<!>},<i!!<u!!aa>},{{{{<!!!>u!!ee!a}a<!!!>!>,<!!!>a!>{!>,<!>},<a!!a>}},{{{<!>,<}{>},{<aaaea!>}a{a!a!>,<!!aa!>,<!!!>>}}}},{<!!!!ae{a<!!!!{!!!!e!>,<,!><!!!>,<>,{}}}},{{{<a!o!!o!!{!>},<>}},{<i!!!!aa<>,<!>},<i!!}}>}},{{<!!{!!!><>,<ea!>},<<!>},<>},{<>,<!>},<<!>!!!>,<!!!>,<>},{<a!><a},!>},<>,{}}}},{{{<!>ea!!a!!ua!!!>>},{<a!!{a!!ei!>},<au>}},{<o!!!!e!><e!>},<{{o<!!!!a>}},{{{{{{<i!!!>{a!>!!!!,a,i,,}!>,<!>},<{!>!>,<>},{<e!!!>o>}},{{<u!!!>!><aioo!!au}!!!>},<!<}>}}},{{<{e!!!!!><!>,<e!!!>aau!!eui>},{{{{{<!a!a!!e}a!!a!!o!!aaoa!>!>},<>}}},<ao!>,<!>},<!!a{iui!>!!!!,!!!><,!>,<a{u>}},{{<o!>},<!!i!uaui!!!!!>!!,,!a!>,<}a!!!>!!>}}}}},{{{{<,!>,<!!!>},<!!!>},<e!>},<!!!>!>,<i!>},<!u{ee>},{<!a!>}>}},{{{<u<!>!!i!>{i!>,<>},<!!{!a!!}ao!>,<!>,<}!>},<!!!<>}}}}},{{{},{<!!!u!!!>!!!oa!!!>},<}!!au!>,<>}},{{},{{<{{<}{!!!>!>e<!,i<{<<!>},<}>},{{<o>}}}}}},{{{<a!!!>!!a!!!>},<}{u>,{{{{<o}!iu{a}!!!>{!>,<!>,<!!!!a!!o!!!>a!>},<!>,<>}},{<!}iao!>},<{!>,<<!>},<!>},<,!>,<,>}},<uaei!!u!}>}},{{<!!e!>,<>},{{<!aue!>i!>,<e>},{<}!oa!!a<{oi!>},<i,!>},<!}!!!>},<i!!>}}},{{},{},{}}},{{},{{{<!!iai!>},<!!!>,!{i<>,{}}},{{{<!!i,!!!!!!!>ei,!>},<!><>}}},{{{}},{{<a!!e!!!!a!>},<,}>,<a}!>},<!!!>!!!>a}!!!!!!<<aaiu>},{{<>}},{{{}},{{<!!a!>,<a!><o!>,<o!!}a!,e!>,o!>,<a!!!>e>}}}},{<a<o},a!>e>,<!>!!!>>}}},{{{{}}},{}},{{<!!aa!!!!i!!!>!!!>!a,>,<a>},{{}}}}}},{{{<}a}!!}ou!!!a!>,<<a}}!!!>u!>i!!!>o>},<a>},{{<!!!>!!}ae!!ei!!!>a!,ue,!>,<!>},<<>},{<!>,<eo!>!>!>!a!!}<!>aa<<i!>,<i!!!>>}},{{<!!a!>},<{!>{!>,<<!>,<,!a{>,<!!!!!>!!{aua!!!!!!!!!>!>},<a!!i<i!!!>!!,!!!>>}}}}},{{{<{!>},<,!}!o!!!>,<!!!>!>u!>,<!!,!!i!!oa{>,<!>,<eua<!>},<o!i<!>},<!a!!!!>},{{<!!!>ou{{aa>},<!>},<{a!!!>a!,}<!!!!a!!a>},{{<!!!>!a!!!<a!!!>},<auai>}}},{{},{{<!>!>!>,<u!{eu<e!>},<<aii!>a!>},<>}},{{{{{{<>}}},<e!>{>},{<a<e!!!!!>,<!>,<}iaa<>}},{<!>,!!!!},,!>{!>,<!aaae>,<!>},<}a}!>},<!a!u}ao>}}},{{{<,au!>},<a{>}},{}},{{{{<<a!>},<!!!{!>e!!{oa!>},<!>e!!a!aa!!!>o>},{<ue!>,<!>,<a!>a!!<,!>},<>}},{{{<!>},<{!!!!!>,u!!!>},<a!!!i!>!!!>,<>}},{{<!!,a!!ie!>oa!>,u}!!!>!>},<a!!,,o!!>},{{{{{{<!>{!!au!>},<a}o!!<!>,<a!!ae>}}},{}},{<a,{i>}}}}}}}},{{{{<!}a!!!>}<!>,<a!!<a!!!>!>},<i,aa,u,>},<a!>e}!>aa>},{{{<,!>auaiue!>},<e}ai>},{}}},{{<!>},<{>},<eiauua!>,<a}!!!>e,oi>}}}},{{{{<!>,<!!!>a!>,<a}{{<oa!!},u{a>}},{{<}!!!ueaa{>,{{<aeoe!!!>>},{{<!a>}}}},{{<{!>,<{>}},{{<aaa<i!{,!>,<,e!>},<o!!!>!!!>!><,!i>},{<e>,<a<!!!>!>,<eu!>!>!!!>e!>!><!!!>i>}}}},{{},{{{{{<<,o!!aa!!a<a!>,<{e}>,{<}>}}}},{{<i{>},<i}!o{!}a,>},{{<a<!>a!oo!a!!<!!ea<!>,<>}}},{{},{{{<!>,<i}!!!>,<!!u{!!!!!>a!!!>e,ae!>},<!>},<!!o>},{<o>}},{{<!>,!!e!!a!>},<}<a<<>},{<o!!e!<!!!>aei!>},<!!{!!!!!>>}},{{{{}},{<!!{oi!>!!eao>}}}},{{{<a!>!!!!,ou!ia!!!>uu!>,<,>}},<!!!>a<!>},<o<!!!aa<u!!!>!!oi}!>,<o>}},{{{{<!!!>u{!!u}!>},<!>},<a!!!>e<!!ai}iea!!!>>},<!i,ia!!oau!!!>a!!,}!>,<!!!>>},{<>,{{<a!>},<a!>,<o{!>,<!>,<i!uu>},{{<aiai!eaio}>},<!>,<{i!>,<a}!!}i!!!!!u!a!>},<!!!>o>}}},{{<,ii>},{{<!>},<!>,<,a!!a>,{<!!!>!!!a!>},<!}{!>aa!>},<eo!>},<!}<o<!o!!!!!!>}},{{<,}a!>!!!>!!!>,<aa!>},<!!!>!aueu!>>},<!!!!>},{<u!>,<!>aa!!!!ia>}},{{{<!!!>!>,<!!a<eu!>,<}!!!!!>o>}}}}}},{{{{<!>},<!}!!u,{,<aie<aa!a!>aa>,{<!>,<,!>},<aoe!>!!!>,>}},{{<>,<!au,o>}},{<!u{>}},{{<e!>iuu{o!!{a>}},{{<<euu!>,<>},{{<!!eau>}}}},{<<>,<<>},{<{>}}},{{<!!o}o<>,{}},{{<}>,{<!!{a!!!>o!>},<e!>},<iu!!e!>,<}!>},<a>,<u!>,<!!!>aa!!!!!>ii{{!>},<!>},<aa,>}},{{{{},{},{{<!u!!!>}a{!!ai!!!>!!!!!!{!!!>>}}},{{<!>,<{o!!!>u<!!{!>ia!>,<oa!>},<>,{<o!!!>}<!!!>},<aee!oua!o{e,uei!!>}}}}}},{<o,,!>,<!!!<a!>e!}!>>,{<!!!>aa!!>,{<{!!,{!>a!!,oiu}{{o!!i>}}}},{{{{{<!>},<a<!>u},aa!,>}},{}},{{<!>},<eaua{o!!i>,{}},<i!>},<!!!!aiu!!>}},{{<eae}i<!>,<>,<e,u!!!!!<!!ea!!u!!!>,<!>,<io!!!>>},{{<!!aua!>},<!!!>!>}!!}<!!a<!!a>},{{<!!!>,<>},<a>}},{<!>,<<a!>,<!au!!!><eu{,!>,<!>},<!!!>}!!!>}!>,<>}},{{{<{,ui!!!!!>},<!!u!>,<!!!>>},{<i!!!!!>!!!>a!!!>!!a},}!>,<<!!!>ua!>,<!>},<>}},{{<a}!>!a!>e!!!>},<}!!!>},<o<aoa>}}},{{<>},{{{<uu{o!>,<i<>}}},{{<iue!>,<!>},<a!!!>!>},<i!!>},{<u!!!>i!>},<!>,<!!!!!>!!!>!!ai<!!!>},<ua<o!<>}}}}}}}"

def group_finder(s):

    openers_stack = []
    group_count = 0
    garbage_count = 0
    bang_ignore = False
    in_bracket_group = False

    for c in s:
        if bang_ignore:
            bang_ignore = False
            continue
        if c == "!":
            bang_ignore = True
            continue

        if in_bracket_group:
            if c == ">":
                in_bracket_group = False
            else:
                garbage_count += 1
            continue
        if c == "<":
            in_bracket_group = True
            continue

        if c == "{":
            openers_stack.append(c)
        if c == "}":
            if not openers_stack:
                return False
            else:
                last_unclosed_opener = openers_stack.pop()
                group_count += len(openers_stack) + 1


    return garbage_count

# assert(group_finder("{}") == 1)
# assert(group_finder("{{{}}}") == 6)
# assert(group_finder("{{}, {}}") == 5)
# assert(group_finder("{<a>,<a>,<a>,<a>}") == 1)
# assert(group_finder("{{<ab>},{<ab>},{<ab>},{<ab>}}") == 9)
# assert(group_finder("{{<!!>},{<!!>},{<!!>},{<!!>}}") == 9)
# assert(group_finder("{{<a!>},{<a!>},{<a!>},{<ab>}}") == 3)

assert(group_finder("<>") == 0)
assert(group_finder("<random characters>") == 17)

print(group_finder(puzzle))
